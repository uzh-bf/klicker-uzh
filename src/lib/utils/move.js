function getIndex(blocks, droppableId) {
  return blocks.findIndex(block => block.id === droppableId)
}

function moveQuestion(blocks, srcBlockId, srcQuestionIx, dstBlockId, dstQuestionIx, removeEmpty = false) {
  // compute the index for both source and destination block
  const srcBlockIx = getIndex(blocks, srcBlockId)
  const dstBlockIx = getIndex(blocks, dstBlockId)

  console.log(blocks.toJS(), srcBlockIx, srcQuestionIx, dstBlockIx, dstQuestionIx)

  // save and remove the question that is to be moved
  const targetQuestion = blocks.getIn([srcBlockIx, 'questions', srcQuestionIx])
  let blocksWithoutSrc = blocks.deleteIn([srcBlockIx, 'questions', srcQuestionIx])

  console.log(blocksWithoutSrc.toJS())

  // if the source block is empty because ot the move, remove it entirely
  // expect if source and destination are the same!
  if (removeEmpty && srcBlockId !== dstBlockId && blocksWithoutSrc.getIn([srcBlockIx, 'questions']).size === 0) {
    blocksWithoutSrc = blocksWithoutSrc.delete(srcBlockIx)
  }

  // compute the new list of questions for the target block
  const dstQuestionsWithTarget = blocksWithoutSrc.getIn([dstBlockIx, 'questions']).insert(dstQuestionIx, targetQuestion)

  // update the target block
  return blocksWithoutSrc.setIn([dstBlockIx, 'questions'], dstQuestionsWithTarget)
}

export { moveQuestion }
