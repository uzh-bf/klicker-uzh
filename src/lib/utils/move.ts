import UUIDv4 from 'uuid/v4'
import _get from 'lodash/get'

export function insertArrayElement(array: any[], index: number, value: any, replace = false): any[] {
  return [...array.slice(0, index), value, ...array.slice(replace ? index + 1 : index)]
}

export function updateArrayElement(array: any[], index: number, newValue: any): any[] {
  const oldValue = array[index]

  if (typeof oldValue === 'object') {
    return insertArrayElement(array, index, { ...oldValue, ...newValue }, true)
  }

  return insertArrayElement(array, index, newValue, true)
}

export function deleteArrayElement(array: any[], index: number): any[] {
  return [...array.slice(0, index), ...array.slice(index + 1)]
}

export function reorder(list: any[], startIndex: number, endIndex: number): any[] {
  const result = Array.from(list)
  const [removed] = result.splice(startIndex, 1)
  result.splice(endIndex, 0, removed)
  return result
}

export function handleDragEnd(array: any, onChange: Function): Function {
  return (result: any): void => {
    if (!result.destination) {
      return
    }

    if (result.destination.index === result.source.index) {
      return
    }

    onChange(reorder(array, result.source.index, result.destination.index))
  }
}

/**
 * Compute the index of a block in a list given its id
 */
export function getIndex(blocks: any[], droppableId: string): number {
  return blocks.findIndex((block): boolean => block.id === droppableId)
}

/**
 * Extend a block with a new question
 * Either at a specific index or by appending to the end
 */
export function addToBlock(blocks: any[], blockId: string | number, question: any, targetIndex: number = null): any[] {
  let dstBlockIx: string | number = blockId

  // if the blockId passed is not number (index)
  // assume it can be calculated by searching for the id
  if (typeof blockId !== 'number') {
    dstBlockIx = getIndex(blocks, blockId)
  }

  // compute the new list of questions for the target block
  const targetValue = { ...question, key: UUIDv4() }
  const dstQuestions: any[] = [..._get(blocks, `${dstBlockIx}.questions`)]
  const dstQuestionsWithTarget =
    typeof targetIndex !== 'undefined' && targetIndex !== null
      ? insertArrayElement(dstQuestions, targetIndex, targetValue)
      : [...dstQuestions, targetValue]

  // update the target block
  const newBlocks = [...blocks]
  newBlocks[dstBlockIx].questions = dstQuestionsWithTarget
  return newBlocks
}

/**
 * Append an entirely new block for a new question
 */
export function appendNewBlock(blocks: any[], question: any): any[] {
  return [
    ...blocks,
    {
      id: UUIDv4(),
      questions: [{ ...question, key: UUIDv4() }],
    },
  ]
}

/**
 * Remove a question from a block
 */
export function removeQuestion(blocks: any[], blockIndex: number, questionIndex: number, removeEmpty = false): any[] {
  // delete the question with the specified index from the specified block
  const blocksWithoutQuestion = [...blocks]
  blocksWithoutQuestion[blockIndex].questions = deleteArrayElement(
    _get(blocks, `${blockIndex}.questions`),
    questionIndex
  )

  // if the block from which the question was removed is now empty, remove the block
  if (removeEmpty && _get(blocksWithoutQuestion, `${blockIndex}.questions`).length === 0) {
    return deleteArrayElement(blocksWithoutQuestion, blockIndex)
  }

  return blocksWithoutQuestion
}

/**
 * Move an existing question from one list to another list
 * @param {*} removeEmpty Flag that specifies whether empty blocks should be pruned
 */
export function moveQuestion(
  blocks: any[],
  srcBlockId: string,
  srcQuestionIx: number,
  dstBlockId: string,
  dstQuestionIx: number,
  removeEmpty = false
): any[] {
  console.log(blocks, srcBlockId, srcQuestionIx, dstBlockId, dstQuestionIx)

  // compute the index for both source and destination block
  const srcBlockIx = getIndex(blocks, srcBlockId)
  let dstBlockIx = getIndex(blocks, dstBlockId)

  // save and remove the question that is to be moved
  const targetQuestion = _get(blocks, `${srcBlockIx}.questions.${srcQuestionIx}`)
  const blocksWithoutSrc = removeQuestion(blocks, srcBlockIx, srcQuestionIx, removeEmpty && srcBlockId !== dstBlockId)

  // if the destination block comes after the source block
  // the destination index needs to be updated after the removal of the source block
  // as everything shifts by one block
  if (blocks.length > blocksWithoutSrc.length && dstBlockIx > srcBlockIx) {
    dstBlockIx -= 1
  }

  return addToBlock(blocksWithoutSrc, dstBlockIx, targetQuestion, dstQuestionIx)
}
