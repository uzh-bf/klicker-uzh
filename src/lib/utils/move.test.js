import { List } from 'immutable'

import { moveQuestion, getIndex } from './move'

describe('move', () => {
  it('can reorder', () => {
    const blocks = List([
      {
        id: 'ba',
        questions: List([
          { id: 'qa', content: 'bla' },
          { id: 'qb', content: 'blu' },
          { id: 'qc', content: 'blee' },
          { id: 'qd', content: 'blii' },
        ]),
      },
    ])
    const reordered = moveQuestion(blocks, 'ba', 1, 'ba', 3)
    expect(reordered.get(0).questions.size).toEqual(4)
    expect(reordered).toMatchSnapshot()
  })

  it('can move', () => {
    const blocks = List([
      {
        id: 'ba',
        questions: List([
          { id: 'qa', content: 'bla' },
          { id: 'qb', content: 'blu' },
          { id: 'qc', content: 'blab' },
          { id: 'qd', content: 'bli' },
        ]),
      },
      {
        id: 'bb',
        questions: List([
          { id: 'qe', content: 'bleb' },
          { id: 'qf', content: 'blaa' },
          { id: 'qg', content: 'blaaa' },
        ]),
      },
      {
        id: 'bc',
        questions: List([
          { id: 'qh', content: 'blaaaa' },
          { id: 'qi', content: 'blabb' },
        ]),
      },
      {
        id: 'bd',
        question: List([{ id: 'qj', content: 'blub' }]),
      },
    ])

    expect(blocks).toMatchSnapshot('#0 initial state')

    // evaluate question movements
    const moved = moveQuestion(blocks, 'bc', 1, 'ba', 3)
    expect(moved).toMatchSnapshot('#1 move c-1 to a-3')

    const moved2 = moveQuestion(moved, 'bc', 0, 'ba', 0, true)
    expect(moved2).toMatchSnapshot('#2 move c-0 to a-0')

    const moved3 = moveQuestion(moved2, 'bb', 1, 'ba', 0)
    expect(moved3).toMatchSnapshot('#3 move b-1 to c-0')

    const moved4 = moveQuestion(moved3, 'bd', 0, 'ba', 0)
    expect(moved4).toMatchSnapshot('#4 move d-0 to a-0')
  })
})
