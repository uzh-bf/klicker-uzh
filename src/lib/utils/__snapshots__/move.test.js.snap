// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`move can move: #0 initial state 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
      Object {
        "content": "blab",
        "id": "qc",
      },
      Object {
        "content": "bli",
        "id": "qd",
      },
    ],
  },
  Object {
    "id": "bb",
    "questions": Immutable.List [
      Object {
        "content": "bleb",
        "id": "qe",
      },
      Object {
        "content": "blaa",
        "id": "qf",
      },
      Object {
        "content": "blaaa",
        "id": "qg",
      },
    ],
  },
  Object {
    "id": "bc",
    "questions": Immutable.List [
      Object {
        "content": "blaaaa",
        "id": "qh",
      },
      Object {
        "content": "blabb",
        "id": "qi",
      },
    ],
  },
  Object {
    "id": "bd",
    "question": Immutable.List [
      Object {
        "content": "blub",
        "id": "qj",
      },
    ],
  },
]
`;

exports[`move can move: #1 move c-1 to a-3 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
      Object {
        "content": "blab",
        "id": "qc",
      },
      Object {
        "content": "blabb",
        "id": "qi",
      },
      Object {
        "content": "bli",
        "id": "qd",
      },
    ],
  },
  Object {
    "id": "bb",
    "questions": Immutable.List [
      Object {
        "content": "bleb",
        "id": "qe",
      },
      Object {
        "content": "blaa",
        "id": "qf",
      },
      Object {
        "content": "blaaa",
        "id": "qg",
      },
    ],
  },
  Object {
    "id": "bc",
    "questions": Immutable.List [
      Object {
        "content": "blaaaa",
        "id": "qh",
      },
    ],
  },
  Object {
    "id": "bd",
    "question": Immutable.List [
      Object {
        "content": "blub",
        "id": "qj",
      },
    ],
  },
]
`;

exports[`move can move: #2 move c-0 to a-0 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      Object {
        "content": "blaaaa",
        "id": "qh",
      },
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
      Object {
        "content": "blab",
        "id": "qc",
      },
      Object {
        "content": "blabb",
        "id": "qi",
      },
      Object {
        "content": "bli",
        "id": "qd",
      },
    ],
  },
  Object {
    "id": "bb",
    "questions": Immutable.List [
      Object {
        "content": "bleb",
        "id": "qe",
      },
      Object {
        "content": "blaa",
        "id": "qf",
      },
      Object {
        "content": "blaaa",
        "id": "qg",
      },
    ],
  },
  Object {
    "id": "bd",
    "question": Immutable.List [
      Object {
        "content": "blub",
        "id": "qj",
      },
    ],
  },
]
`;

exports[`move can move: #3 move b-1 to c-0 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      Object {
        "content": "blaa",
        "id": "qf",
      },
      Object {
        "content": "blaaaa",
        "id": "qh",
      },
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
      Object {
        "content": "blab",
        "id": "qc",
      },
      Object {
        "content": "blabb",
        "id": "qi",
      },
      Object {
        "content": "bli",
        "id": "qd",
      },
    ],
  },
  Object {
    "id": "bb",
    "questions": Immutable.List [
      Object {
        "content": "bleb",
        "id": "qe",
      },
      Object {
        "content": "blaaa",
        "id": "qg",
      },
    ],
  },
  Object {
    "id": "bd",
    "question": Immutable.List [
      Object {
        "content": "blub",
        "id": "qj",
      },
    ],
  },
]
`;

exports[`move can move: #4 move d-0 to a-0 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      undefined,
      Object {
        "content": "blaa",
        "id": "qf",
      },
      Object {
        "content": "blaaaa",
        "id": "qh",
      },
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
      Object {
        "content": "blab",
        "id": "qc",
      },
      Object {
        "content": "blabb",
        "id": "qi",
      },
      Object {
        "content": "bli",
        "id": "qd",
      },
    ],
  },
  Object {
    "id": "bb",
    "questions": Immutable.List [
      Object {
        "content": "bleb",
        "id": "qe",
      },
      Object {
        "content": "blaaa",
        "id": "qg",
      },
    ],
  },
  Object {
    "id": "bd",
    "question": Immutable.List [
      Object {
        "content": "blub",
        "id": "qj",
      },
    ],
  },
]
`;

exports[`move can reorder 1`] = `
Immutable.List [
  Object {
    "id": "ba",
    "questions": Immutable.List [
      Object {
        "content": "bla",
        "id": "qa",
      },
      Object {
        "content": "blee",
        "id": "qc",
      },
      Object {
        "content": "blii",
        "id": "qd",
      },
      Object {
        "content": "blu",
        "id": "qb",
      },
    ],
  },
]
`;
